src/Application/ChangedNodeScopeRefresher.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Core\Application;
+
+use PhpParser\Node;
+use PhpParser\Node\Attribute;
+use PhpParser\Node\AttributeGroup;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Identifier;
+use PhpParser\Node\Param;
+use PhpParser\Node\Stmt;
+use PhpParser\Node\Stmt\Enum_;
+use PhpParser\Node\Stmt\Expression;
+use PhpParser\Node\Stmt\Property;
+use PHPStan\Analyser\MutatingScope;
+use Rector\Core\Exception\ShouldNotHappenException;
+use Rector\NodeTypeResolver\PHPStan\Scope\PHPStanNodeScopeResolver;
+use Symplify\SmartFileSystem\SmartFileInfo;
+
+/**
+ * In case of changed node, we need to re-traverse the PHPStan Scope to make all the new nodes aware of what is going on.
+ */
+final class ChangedNodeScopeRefresher
+{
+    public function __construct(
+        private readonly PHPStanNodeScopeResolver $phpStanNodeScopeResolver,
+    ) {
+    }
+
+    public function refresh(
+        Expr|Stmt|Node $node,
+        SmartFileInfo $smartFileInfo,
+        MutatingScope $mutatingScope
+    ): void {
+        // nothing to refresh
+        if ($node instanceof Identifier) {
+            return;
+        }
+
+        // note from flight: when we traverse ClassMethod, the scope must be already in Class_, otherwise it crashes
+        // so we need to somehow get a parent scope that is already in the same place the $node is
+
+        if ($node instanceof Attribute) {
+            // we'll have to fake-traverse 2 layers up, as PHPStan skips Scope for AttributeGroups and consequently Attributes
+            $attributeGroup = new AttributeGroup([$node]);
+            $node = new Property(0, [], [], null, [$attributeGroup]);
+        }
+
+        // phpstan cannot process for some reason
+        if ($node instanceof Enum_) {
+            return;
+        }
+
+        if ($node instanceof Stmt) {
+            $stmts = [$node];
+        } elseif ($node instanceof Expr) {
+            $stmts = [new Expression($node)];
+        } else {
+            if ($node instanceof Param) {
+                // param type cannot be refreshed
+                return;
+            }
+
```

Comment:
How about `Arg` ?

Reply:
So far args have scope, as they're `Expr`.

Reply:
Arg is extends NodeAbstract directly, how can Arg be Expr ? https://github.com/nikic/PHP-Parser/blob/a6e34665fd6a2bcefc924cb5fd73788767ae510e/lib/PhpParser/Node/Arg.php#L8

Reply:
Ah, I see. I meant `$arg->value`. So far the rules do not hook on it as the context is too shallow.
So not neccassary to replicate scope to all args. It will hold more memory to refer scope, so the usage of these must have some valid reason.

Reply:
Then, Arg should return early:

```php
if ($node instanceof Arg) {
    return;
}
```

to avoid throw Exception